### DISCUSSION TITLE
Performance View

### DISCUSSION THREAD

**Discussion by entzmingerc** [SynthstromAudible/DelugeFirmware#712] [Ideas]
This is just information from: 11/11/2023, everything is subject to change in future (SOON)

feature author: @seangoodvibes 
description: Each column of the main grid corresponds to a Song View level FX parameter. Each row of the main grid corresponds to a value of the Song View level FX parameter. The intention is to play the FX similar to other grid based performance FX units (polyend tracker performance mode, roland scatter FX, others)

usage:
1) Enter performance view from song view / grid view by pressing keyboard button
2) Short press a pad in the columns to hold that pad's value (parameter and held value is displayed on the screen)
3) Long press a pad in the columns to play the FX but not hold it (when you let go of the pad it will snap back to previous value). 

Columns, Colors, FX
1 = Red = LPF Cutoff 
2 = Red = LPF Resonance
3 = Orange = HPF Cutoff
4 = Orange = HPF Resonance
5 = Yellow = Bass EQ
6 = Yellow = Treble EQ
7 = Light Green = Reverb
8 = Light Blue = Delay Amount
9 = Light Blue = Delay Rate
10 = Light Pink = MOD-FX Offset
11 = Light Pink = MOD-FX Feedback
12 = Light Pink = MOD-FX Depth
13 = Light Pink = MOD-FX Rate
14 = Dark Pink = Decimation
15 = Dark Pink = Bitcrush
16 = Dark Blue = Stutter Rate + Stutter Trigger

Row values: 0, 7, 14, 21, 28, 35, 42, 50

Warning! Delay amount will erupt into loud feedback past values of 25 (as is typical with the deluge digital delay). Switch to analog delay mode for a less immediate feedbacking :)

Warning! Don't download random binaries on the internet.
However, there is a random binary on the internet on the discord if you'd like to try it out.

**Comment by entzmingerc** [comment]
### Feedback and requests submitted so far

- Swtich to performance view as a grid mode, use the color Pink for it. 
- Be able to select what parameter each column is set to. Possibly use [select turn] somehow to browse FX. 
- Clear all FX settings with [BACK] + [press <>] to keep it aligned with other Deluge Views.
- Use [turn <>] to rotate the columns. Intention: press a pad, then rotate columns, it should set and hold the values of each of the columns you rotate through. Multiple pad presses and then rotate should set all the columns you're pressing?
- A more complex idea was to set each column to control multiple parameters somehow to develop a "macro". It might be possible to reuse the Automation View Overlay for a column to show what parameters a column is mapped to, and for what modulation depth. 
- Slew time for parameter changes could provide an interesting way to change values, instead of instantaneously changing parameter values when pressing a pad. (Similar to a "portamento" or "interpolation" style of value changes).
- If stutter is active, it locks users out of switching views, which is consistent with official firmware(?).
- The ability to set row values of pads for each column has been requested. Loud feedback at values above 25 for digital delay, which is desirable to some and dangerous to others. Setting pad values to specific values is useful for delay amount, fine tuning delay time to different related values, fine tuning LPF and HPF cutoffs to nuanced filter settings, fine tuning resonance so that it does / doesn't self-oscillate the filters, and so on. 
- If row values can be set to specific values, how do we save them? Could be tied to the song. Could be tied to a preset similar to a synth or kit preset so that we load up different performance view presets, otherwise we'd have to remake our performance view "template" from scratch. 
- A more advanced idea brought up would be to sequence the templates. "Set a bunch of values, remove all FX, then re-enable the same FX". Switching between presets of FX. 
- If you move the gold knob of another param the stutter stops.
- "Make delay fade out when deselecting it." (I personally think you shouldn't do this because it's inconsistent behavior from the other params and not how delay amount works since it controls both the level and the feedback) but could be cool. If anything, this indicates decoupling the delay level and delay feedback would be a beneficial thing in general. 
- "Select which tracks are affected." This sounds inconsistent with expectations since these are Song View level FX which have always affected everything, but it would provide more flexibility. Something like a Dry/Wet mix that's defaulted to 100% for each clip. Unsure about audio code pathing and the UX for this.
- BPM sync'ing FX. Look at LFO or sidechain code for example, calculate a rate that takes 1/4 note of ticks to complete and then half/double it to sync to different tempos. 

("view" or "mode"?)

**Reply by duncanrmi** [reply]
one of the few things I like about the polyend play is this mode. so, yeah.... bring it on!

**Reply by entzmingerc** [reply]
Sean has already outdated this list! :)
Clear all FX works, we can set each pad in each column to a specific value and save the preset, we can set each column to a different FX, there's an overview of the available FX so far, and loading presets I think is working.

**Comment by entzmingerc** [comment]
![image](https://github.com/SynthstromAudible/DelugeFirmware/assets/87605057/cc7ecf6b-463d-4412-869f-f13be68f1763)

**Reply by AtraWanderer** [reply]
It would be incredible to still have the clips on/of on the rigth to still be able to change them no?

**Comment by seangoodvibes** [comment]
> It would be incredible to still have the clips on/of on the rigth to still be able to change them no?

Yes I'll work on that.
